src/main/java/org/jetbrains/bsp/bazel/server/BepServer.java

```diff
@@ -88,156 +87,122 @@ public void publishLifecycleEvent(
-              Position position =
-                  new Position(
-                      Integer.parseInt(lineLocation[0]), Integer.parseInt(lineLocation[1]));
-              Diagnostic diagnostic = new Diagnostic(new Range(position, position), error);
-              diagnostic.setSeverity(DiagnosticSeverity.ERROR);
-              List<Diagnostic> diagnostics =
-                  fileDiagnostics.getOrDefault(fileLocation, new ArrayList<>());
-              diagnostics.add(diagnostic);
-              fileDiagnostics.put(fileLocation, diagnostics);
-            });
+    StatusCode exitCode = BepServer.convertExitCode(buildFinished.getExitCode().getCode());
+    TaskFinishParams finishParams = new TaskFinishParams(taskParkingLot.pop(), exitCode);
+    finishParams.setEventTime(buildFinished.getFinishTimeMillis());
 
-    fileDiagnostics.forEach(
-        (fileLocation, diagnostics) ->
-            bspClient.onBuildPublishDiagnostics(
-                new PublishDiagnosticsParams(
-                    new TextDocumentIdentifier(Uri.fromAbsolutePath(fileLocation).toString()),
-                    new BuildTargetIdentifier(""),
-                    diagnostics,
-                    false)));
+    bspClient.onBuildTaskFinish(finishParams);
   }
 
-  private void processCompletionEvent(BuildEventStreamProtos.BuildEvent event) throws IOException {
-    List<BuildEventStreamProtos.OutputGroup> outputGroups =
-        event.getCompleted().getOutputGroupList();
+  private void processProgressEvent(BuildEventStreamProtos.Progress progress) {
+    HashMap<String, List<Diagnostic>> fileDiagnostics = parseStdErrDiagnostics(progress);
+
+    for (Entry<String, List<Diagnostic>> entry : fileDiagnostics.entrySet()) {
+      String fileLocation = entry.getKey();
+      List<Diagnostic> diagnostics = entry.getValue();
+
+      PublishDiagnosticsParams publishDiagnosticsParams =
+          new PublishDiagnosticsParams(
+              new TextDocumentIdentifier(Uri.fromAbsolutePath(fileLocation).toString()),
+              new BuildTargetIdentifier(""),
+              diagnostics,
+              false);
+
+      bspClient.onBuildPublishDiagnostics(publishDiagnosticsParams);
+    }
+
+    buildClientLogger.logMessage(progress.getStderr().trim());
+  }
+
+  private void processCompletedEvent(BuildEventStreamProtos.TargetComplete targetComplete)
+      throws IOException {
+    List<OutputGroup> outputGroups = targetComplete.getOutputGroupList();
+
     if (outputGroups.size() == 1) {
       BuildEventStreamProtos.OutputGroup outputGroup = outputGroups.get(0);
-      if ("scala_compiler_classpath_files".equals(outputGroup.getName())) {
+
+      if (outputGroup.getName().equals("scala_compiler_classpath_files")) {
         for (BuildEventStreamProtos.BuildEventId.NamedSetOfFilesId fileSetId :
             outputGroup.getFileSetsList()) {
           for (BuildEventStreamProtos.File file :
```

Comment:
nested fors should be extracted to the methods

Reply:
Yeah, I know, there's a lot to extract there, but I left those 3 methods because they're too messy and in many places I know that a method should be extracted but I don't know what this piece of code does precisely and I can't find names for them xd

Reply:
cool, tomorrow I will try to look at this part and I will share my thoughts and ideas for naming